Extend list widget with "index" attribute

Jermolene opened the pull request:
This PR extends the `<$list>` widget with an optional `index` parameter that specifies the name of a variable to receive the numeric index of each list item. For example:

```
<$list filter="[tag[About]sort[title]]" index="index">
<div>
<<index>>: ''<$text text=<<currentTiddler>>/>'' (is first: <<index-first>>, is last: <<index-last>>)
</div>
</$list>
```

Two additional variables are also set to indicate the first and last items in the list:

* `<index-variable-name>-first` is set to `yes` for the first entry in the list, `no` for the others
* `<index-variable-name>-last` is set to `yes` for the last entry in the list, `no` for the others

This PR fixes #1330 and #3384. It probably obviates #1523. It also finally replaces the reverted PR #1328 from 2015.

saqimtiaz commented:
@Jermolene this looks good and I favour this approach rather than introducing a new widget as has been previously discussed. 

We may want to draw more attention to the note about degrading performance, but we can fine tune the docs once this is merged.

Jermolene commented:
> @Jermolene this looks good and I favour this approach rather than introducing a new widget as has been previously discussed.

The result is a bit cleaner than I feared; the list widget is still a bit of a mess, but not too much worse!

pmario commented on core/modules/widgets/list.js:
```diff
@@ -128,7 +129,14 @@ ListWidget.prototype.makeItemTemplate = function(title) {
+	if(this.indexName) {
+		parseTreeNode.index = index.toString();
+		parseTreeNode.indexName = this.indexName;
+		parseTreeNode.isFirst = index === 0;
```
hmm, Should we use 0-based counting for user-faced numbers? eg [nth operator](https://tiddlywiki.com/#nth%20Operator) starts with 1  ?!

Jermolene replied:
The `reduce[]` operator uses zero-based indexes, and I think it's reasonable to continue. (The `nth` operator was part of one of those PRs that I've come to regret)

pmario replied:
Then we should deprecate it and remove it in a year or so.

kookma commented:
> We may want to draw more attention to the note about degrading performance, but we can fine tune the docs once this is merged.

$list is an essential widget and I hope we have no degrading performance!

saqimtiaz commented:
@kookma the performance degradation only applies if you use the new index parameter (in which case there is no way around it). Otherwise the performance will be the same.

Jermolene commented:
An interesting experiment to assess the performance difference between the two variants of the list widget is to hack the code to always use the index attribute. I did it by splicing the following line in:

```js
	this.indexName = this.getAttribute("index");
	this.indexName = this.indexName || "unused-index"; // <<<<< ADD THIS LINE
	this.storyViewName = this.getAttribute("storyview");
```

Playing with the resulting wiki, as you'd expect the story river becomes very sluggish opening and closing tiddlers. The tab controls behave hilariously: every tab "pops" as you switch between them.

I uploaded a snapshot here for others to experiment with:

https://tw5-with-hacked-list-widget.tiddlyhost.com

pmario commented on core/modules/widgets/list.js:
```diff
@@ -142,7 +150,7 @@ ListWidget.prototype.refresh = function(changedTiddlers) {
 	}
 	// Completely refresh if any of our attributes have changed
-	if(changedAttributes.filter || changedAttributes.template || changedAttributes.editTemplate || changedAttributes.emptyMessage || changedAttributes.storyview || changedAttributes.history) {
+	if(changedAttributes.filter || changedAttributes.variable || changedAttributes.index || changedAttributes.template || changedAttributes.editTemplate || changedAttributes.emptyMessage || changedAttributes.storyview || changedAttributes.history) {
```
IMO if there is a refresh needed, if any attribute changes, the mechanism that Saq uses is much more readable.

```
if (changedAttributes.length > 0) {
```

Jermolene commented:
Just to note that in 4a99e0cc7d4a6b9e7071c0b2a9a0f63c3c7d2492, I changed the name of the "index" attribute to "counter", and moved to 1-based counting.
